RELEASES.md

```diff
@@ -205,3 +205,95 @@ To create a minor release:
+process at specific pre-defined points in its execution so that we can verify
+upgrades work in a representative sample of stop conditions.
+
+### Large Scale Testnets
+
+The Tendermint end-to-end tests run a small network (~10s of nodes) to exercise
+basic consensus interactions. Real world deployments of Tendermint often have over 
+a hundred nodes just in the validator set, with many others acting as full
+nodes and sentry nodes. To gain more assurance before a release, we will also run
+larger-scale test networks to shake out emergent behaviors at scale.
+
+Large-scale test networks are run on a set of virtual machines (VMs). Each VM
+is equipped with 4 Gigabytes of RAM and 2 CPU cores. The network runs a very
+simple key-value store application. During each test net, the following metrics
+are monitored and collected on each node:
+
+* Tendermint rounds per height
+* Peers connected
+* Memory resident set size
+* CPU utilization
+* Blocks produced per minute
+* Seconds for each step of consensus (Propose, Prevote, Precommit, Commit)
+* Latency to receive each block proposal
+
+For these tests we intentionally target low-powered host machines (with low core
+counts and limited memory) to ensure we observe similar kinds of resource contention 
+and limitation that real-world  deployments of Tendermint experience in production. 
+
+#### 200 Node Testnet
+
+To test the stability and performance of Tendermint in a real world scenario,
+a 200 node test network is run. The network comprises 5 seed nodes, 100
+validators and 95 non-validating full nodes. All nodes begin by dialing
+a subset of the seed nodes to discover peers. The network is run for several
+days, with metrics being collected continuously.
+ 
+#### Rotating Node Testnet
+
+Real-world deployments of Tendermint frequently see new nodes arrive and old
+nodes exit the network. The rotating node testnet ensures that Tendermint is
+able to handle this reliably. In this test, a network with 10 validators and
+3 seed nodes is started. A rolling set of 25 full nodes are started and each
+connects to the network by dialing one of the seed nodes. Once the node is able
+to blocksync to the head of the chain and begins producing blocks using
+Tendermint consensus it is stopped. Once stopped, a new node is started and
+takes its place. This network is run for several days.
+
+#### Absent Stake Testnet
+
+Tendermint networks often run with _some_ portion of the voting power offline.
+The absent stake testnet ensures that large networks are able to handle this
+reliably. A set of 150 validator nodes and three seed nodes is started.
+The set of 150 validators is configured to only possess a cumulative stake of 67%
+of the total stake. The remaining 33% of the stake is configure to belong to a
+validator that is never actually run in the test network. The network is run
+for multiple days, ensuring that it is able to produce blocks without issue.
```

Comment:
> Partial degradation: losses, increased delay (e.g., using "tc" or similar)

I'm not totally against this, but I'm not sure I see what specific class of issues would be identified by this type of testing. Tendermint will degrade under these conditions, as expected. Are there types of degradation that we expect this type of testing to catch that can't be caught in a simpler way? 

> Network partitions (which cause the chain to halt until it is healed)

I think this might be a better candidate for a simple e2e nightly test, since I'm not sure what will be caught at scale that wouldn't be caught in a simpler and smaller network. If there's a change that causes Tendermint to not behave correctly during partitions it will be due to a logic bug that is likely to be more difficult to catch and diagnose when we need to examine 100 nodes for evidence of the issue. 

The reasoning, imo, for the absent stake scenario is to test that vote and proposal propagation times are performant enough in large networks. With a lower % of stake, fewer (or no) votes can be missing or nil for a block to be committed. If there are degradation in Tendermint's ability to gossip messages, either due to consensus code changes, reactor changes, or p2p changes, they should show up in this kind of test while they are less likely to show up on the small scale.

> [maybe too much?] syn attack

I continue to be of the belief that network and TCP-level attacks should be concerns of the TCP stack and tools that are specifically designed for solving these problems and that this is both not something that we can reasonably test nor is it something we should specialize in. The solution to many of these types of attacks are standard fare, like firewalls w/ rate-limiting, disabling UDP, and turning on SYN-cookies none of which are really Tendermint's job.

Reply:
I acknowledge that some of the problems that may surface with this type of tests could also be seen in _regular_ e2e tests. However:
* with a sizable network, we may stumble upon resource-related problems (e.g., memory, lock or network contention) that may not manifest in a smaller e2e network.
* these perturbations can (and will) definitely happen in the wild. Tendermint is not expected to operate normally _during_ those perturbations, but should continue unaffected once they stop (i.e., only liveness should be affected).

So, what I'm actually proposing is bootstrapping a sizeable network, having it reach normal operation, and then applying (one of) the mentioned perturbations for a few seconds (minutes?), and then verifying that the system resumes normal operation when the perturbations stop.

I prefer to bump into these problems _before_ the release for two reasons: (a) to be able to react earlier, when not all eyes are set on us, (b) in case a bad behavior is unfortunately detected later, in App testnets, or in production, we can show to what extent we went in our testing, as a way to help keep our users' confidence.

I think it is worth it for such a critical part of Cosmos, but that is just my opinion.

Reply:
I agree with Sergio in the sense that this particular test is not testing performance only but correctness as well. We caught some problems in p2p for example when one peer is slow etc.

Reply:
OK, I agree on the "slow and unwieldy" case, that's why this is typically only done once per release. And that's what justifies the "feature freeze" or "code freeze" you describe in the text (you can't be doing this as a CI step).

I'm not sure I fully agree on the part where you say "if we actually do catch an issue with correctness at this stage that it will be so difficult to understand". Yes, it will be difficult and involved, but it's still a problem that will need attention. And, again, I prefer to find this (if possible) before it hits our users. There it'll be as difficult, but on top, we'll have everyone's attention.

Anyway, the stages you have in the current version of the document seem reasonable to me, so if you add the network partition, I'm happy with the result.

Reply:
> Yes, it will be difficult and involved, but it's still a problem that will need attention.

Again, I completely agree. I am advocating that we try to do everything in our ability to catch issues before this stage because of the high cost of fully diagnosing.

> Anyway, the stages you have in the current version of the document seem reasonable to me, so if you add the network partition, I'm happy with the result.

Added a section, please take a look :pray:

Reply:
@williambanfield , I just went through it and looks good to me. Thanks a lot